clang/runtime/dpct-rt/include/math.hpp

```diff
@@ -36,6 +37,44 @@ class vectorized_binary<
 
 template <typename T> bool isnan(const T a) { return sycl::isnan(a); }
 // TODO: Need add more specialization such as bfloat16 version.
+
+template <typename T>
+inline int64_t zeroOrSignedExtent(const T val, const unsigned bit) {
+  if constexpr (std::is_signed_v<T>) {
+    return int64_t(val << (64 - bit)) >> (64 - bit);
+  }
+  return val;
+}
+
+template <typename retT, typename AT, typename BT, class BinaryOperation>
+inline constexpr retT extend_binary(const AT a, const BT b,
+                                    const BinaryOperation binary_op,
+                                    bool need_sat = false) {
+  const auto extend_a = zeroOrSignedExtent(a, 33);
+  const auto extend_b = zeroOrSignedExtent(b, 33);
+  const auto ret = binary_op(extend_a, extend_b);
+  if (need_sat)
+    return sycl::clamp(ret, std::numeric_limits<retT>::min(),
+                       std::numeric_limits<retT>::max());
+  return ret;
+}
+
+template <typename retT, typename AT, typename BT, typename CT,
+          class BinaryOperation>
+inline constexpr retT extend_binary(const AT a, const BT b, const CT c,
+                                    const BinaryOperation binary_op,
+                                    const BinaryOperation second_op,
+                                    bool need_sat = false) {
+  const auto extend_a = zeroOrSignedExtent(a, 33);
+  const auto extend_b = zeroOrSignedExtent(b, 33);
+  const auto extend_temp =
+      zeroOrSignedExtent(binary_op(extend_a, extend_b), 34);
+  if (need_sat)
```

Comment:
Can we avoid use `if`, it may caused performance issue.

Reply:
fixed